pkg/registry/core/service/strategy.go

```diff
@@ -226,3 +226,109 @@ func (serviceStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runt
 func (serviceStatusStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
 	return validation.ValidateServiceStatusUpdate(obj.(*api.Service), old.(*api.Service))
 }
+
+// normalizeClusterIPs adjust clusterIPs based on ClusterIP
+func normalizeClusterIPs(oldSvc *api.Service, newSvc *api.Service) {
+	// In all cases here, we don't need to over-think the inputs.  Validation
+	// will be called on the new object soon enough.  All this needs to do is
+	// try to divine what user meant with these linked fields. The below
+	// is verbosely written for clarity.
+
+	if oldSvc == nil {
+		// This was a create operation.
+		// User specified singular and not plural (e.g. an old client), so init
+		// plural for them.
+		if len(newSvc.Spec.ClusterIP) > 0 && len(newSvc.Spec.ClusterIPs) == 0 {
+			newSvc.Spec.ClusterIPs = []string{newSvc.Spec.ClusterIP}
+			return
+		}
+
+		// User specified plural and not singular (new client) so init singular for them.
+		if len(newSvc.Spec.ClusterIPs) > 0 && len(newSvc.Spec.ClusterIP) == 0 {
+			newSvc.Spec.ClusterIP = newSvc.Spec.ClusterIPs[0]
+			return
+		}
+
+		// Either both were not specified (will be allocated) or both were
+		// specified (will be validated).
+		return
+	}
+
+	// This was an update operation.
+
+	// **** IMPORTANT *****
+	// Old clients operate only on ClusterIP. New Clients
+	// *MUST* operate on both fields (except when setting ClusterIPs for the first time)
+
+	// ClusterIPs were cleared by an old client which was trying to patch
+	// some field and didn't provide ClusterIPs
+	if len(oldSvc.Spec.ClusterIPs) > 0 && len(newSvc.Spec.ClusterIPs) == 0 {
+		// if ClusterIP is the same, then it is an old client trying to
+		// patch service and didn't provide ClusterIPs
+		if oldSvc.Spec.ClusterIP == newSvc.Spec.ClusterIP {
+			newSvc.Spec.ClusterIPs = oldSvc.Spec.ClusterIPs
+		}
```

Comment:
We need to do this for IPFamilyPolicy and IPFamilies as well... or is that handled somewhere else?

Reply:
IPFamilies and IPFamily Policy does not need this. This is really meant to normalize ClusterIP<=>ClusterIPs since they are closely related and have the patching issue (e.g. clear one without clearing the other).